### PR TITLE
[New Azure Integration Pkgs] Add DataProtection blob storage implementation based on latest Azure.Storage.Blobs

### PR DESCRIPTION
**Pull request by BrennanConroy**

### CODE REVIEW

File: src/DataProtection/Azure.Storage.Blob/src/Microsoft.AspNetCore.DataProtection.Azure.Storage.Blob.csproj [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+
+  <PropertyGroup>
+    <Description>Microsoft Azure Blob storage support as key store.</Description>
+    <TargetFramework>netstandard2.0</TargetFramework>
+    <AllowUnsafeBlocks>true</AllowUnsafeBlocks>
+    <GenerateDocumentationFile>true</GenerateDocumentationFile>
+    <PackageTags>aspnetcore;dataprotection;azure;blob</PackageTags>
+    <IsShippingPackage>true</IsShippingPackage>
+    <UseLatestPackageReferences>true</UseLatestPackageReferences>
+    <HasReferenceAssembly>false</HasReferenceAssembly>
```
Comment (dougbu) [inline src/DataProtection/Azure.Storage.Blob/src/Microsoft.AspNetCore.DataProtection.Azure.Storage.Blob.csproj]: nit: Shouldn't need this because `$(IsAspNetCoreApp)` is `false`.

File: src/DataProtection/Azure.Storage.Blob/src/AzureBlobXmlRepository.cs [hunk 2]
```diff
@@ -0,0 +1,291 @@
...
+
+                using (var xmlReader = XmlReader.Create(memoryStream, xmlReaderSettings))
+                {
+                    return XDocument.Load(xmlReader);
+                }
+            }
+        }
+
+        private BlobClient CreateFreshBlobRef()
+        {
```
Comment (pakrym) [inline src/DataProtection/Azure.Storage.Blob/src/AzureBlobXmlRepository.cs]: BlobClient is threadsafe now

File: src/DataProtection/Azure.Storage.Blob/src/AzureBlobXmlRepository.cs [hunk 3]
```diff
@@ -0,0 +1,291 @@
...
+                using (var memoryStream = new MemoryStream())
+                {
+                    await blobRef.DownloadToAsync(
+                        destination: memoryStream,
+                        conditions: requestCondition);
+
+                    // At this point, our original cache either didn't exist or was outdated.
+                    // We'll update it now and return the updated value;
+
+                    var props = await blobRef.GetPropertiesAsync();
```
Comment (pakrym) [inline src/DataProtection/Azure.Storage.Blob/src/AzureBlobXmlRepository.cs]: Method call above returns Response, you can get ETag from headers property on it.

File: src/DataProtection/Azure.Storage.Blob/src/AzureBlobXmlRepository.cs [hunk 4]
```diff
@@ -0,0 +1,291 @@
...
+                serializedDoc.Position = 0;
+
+                // Generate the appropriate precondition header based on whether or not
+                // we believe data already exists in storage.
+
+                BlobRequestConditions requestConditions;
+                BlobHttpHeaders headers = null;
+                if (latestData != null)
+                {
+                    var props = await blobRef.GetPropertiesAsync();
```
Comment (pakrym) [inline src/DataProtection/Azure.Storage.Blob/src/AzureBlobXmlRepository.cs]: We shouldn't be requesting etag here, we need to use the one `_cachedBlobData` was revieved with. Intent is to only update the value if we have the latest.

File: src/DataProtection/Azure.Storage.Blob/src/AzureDataProtectionBuilderExtensions.cs [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+                    message: "URI does not have a SAS token in the query string.",
+                    paramName: nameof(blobUri));
+            }
+
+            var credentials = new StorageSharedKeyCredential(uriBuilder.AccountName, uriBuilder.Query);
+            var blobName = uriBuilder.BlobName;
+            uriBuilder.Query = null; // no longer needed
+            var blobAbsoluteUri = uriBuilder.ToUri();
+
+            var client = new BlobContainerClient(blobAbsoluteUri, credentials);
```
Comment (pakrym) [inline src/DataProtection/Azure.Storage.Blob/src/AzureDataProtectionBuilderExtensions.cs]: Just pass the `blobUri` to `BlobContainerClient` ctor. It should work 🤞

File: src/DataProtection/Azure.Storage.Blob/src/AzureDataProtectionBuilderExtensions.cs [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+        /// in Azure Blob Storage.
+        /// </summary>
+        /// <param name="builder">The builder instance to modify.</param>
+        /// <param name="blobUri">The full URI where the key file should be stored.
+        /// The URI must contain the SAS token as a query string parameter.</param>
+        /// <returns>The value <paramref name="builder"/>.</returns>
+        /// <remarks>
+        /// The container referenced by <paramref name="blobUri"/> must already exist.
+        /// </remarks>
+        public static IDataProtectionBuilder PersistKeysToAzureBlobStorage(this IDataProtectionBuilder builder, Uri blobUri)
```
Comment (pakrym) [inline src/DataProtection/Azure.Storage.Blob/src/AzureDataProtectionBuilderExtensions.cs]: We need an overload that takes uri and TokenCredential

File: src/DataProtection/Azure.Storage.Blob/src/AzureDataProtectionBuilderExtensions.cs (hunk 5 above)
Comment (pakrym) [reply src/DataProtection/Azure.Storage.Blob/src/AzureDataProtectionBuilderExtensions.cs]: I think I'm wrong. What this overload should do is to see if Uri is SAS uri and if not use DefaultAzureCredential for authentication.

File: src/DataProtection/Azure.Storage.Blob/src/AzureDataProtectionBuilderExtensions.cs [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+        /// <param name="builder">The builder instance to modify.</param>
+        /// <param name="container">The <see cref="BlobContainerClient"/> in which the
+        /// key file should be stored.</param>
+        /// <param name="blobName">The name of the key file, generally specified
+        /// as "[subdir/]keys.xml"</param>
+        /// <returns>The value <paramref name="builder"/>.</returns>
+        /// <remarks>
+        /// The container referenced by <paramref name="container"/> must already exist.
+        /// </remarks>
+        public static IDataProtectionBuilder PersistKeysToAzureBlobStorage(this IDataProtectionBuilder builder, BlobContainerClient container, string blobName)
```
Comment (pakrym) [inline src/DataProtection/Azure.Storage.Blob/src/AzureDataProtectionBuilderExtensions.cs]: At this point why not take a `BlobClient`

File: src/DataProtection/Azure.Storage.Blob/src/AzureDataProtectionBuilderExtensions.cs (hunk 5 above)
Comment (pakrym) [reply src/DataProtection/Azure.Storage.Blob/src/AzureDataProtectionBuilderExtensions.cs]: To check if it's a SAS url create a `bloburibuilder` and check the https://docs.microsoft.com/en-us/dotnet/api/azure.storage.blobs.bloburibuilder.sas?view=azure-dotnet#Azure_Storage_Blobs_BlobUriBuilder_Sas property

File: src/DataProtection/Azure.Storage.Blob/src/AzureDataProtectionBuilderExtensions.cs (hunk 7 above)
Comment (BrennanConroy) [reply src/DataProtection/Azure.Storage.Blob/src/AzureDataProtectionBuilderExtensions.cs]: So we should replace this overload?

Comment (pakrym) [reply src/DataProtection/Azure.Storage.Blob/src/AzureDataProtectionBuilderExtensions.cs]: Yep

File: src/DataProtection/Azure.Storage.Blob/src/AzureDataProtectionBuilderExtensions.cs [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+            BlobContainerClient client;
+
+            // The SAS token is present in the query string.
+            if (uriBuilder.Sas == null)
+            {
+                client = new BlobContainerClient(blobUri, new DefaultAzureCredential());
+            }
+            else
+            {
+                var credentials = new StorageSharedKeyCredential(uriBuilder.AccountName, uriBuilder.Query);
```
Comment (pakrym) [inline src/DataProtection/Azure.Storage.Blob/src/AzureDataProtectionBuilderExtensions.cs]: I this branch you can pass the URI directly into the constructor.

File: src/DataProtection/Azure.Storage.Blob/src/AzureDataProtectionBuilderExtensions.cs [hunk 9]
```diff
@@ -0,0 +1,131 @@
...
+                throw new ArgumentNullException(nameof(builder));
+            }
+            if (blobClient == null)
+            {
+                throw new ArgumentNullException(nameof(blobClient));
+            }
+            return PersistKeystoAzureBlobStorageInternal(builder, blobClient);
+        }
+
+        // important: the Func passed into this method must return a new instance with each call
```
Comment (pakrym) [inline src/DataProtection/Azure.Storage.Blob/src/AzureDataProtectionBuilderExtensions.cs]: Remove the comment

File: src/DataProtection/Azure.Storage.Blob/src/Microsoft.AspNetCore.DataProtection.Azure.Storage.Blob.csproj [hunk 10]
```diff
@@ -0,0 +1,25 @@
...
+    <DotNetFinalVersionKind />
+  </PropertyGroup>
+
+  <ItemGroup>
+    <Reference Include="Microsoft.AspNetCore.DataProtection" />
+    <Reference Include="Azure.Storage.Blobs" />
+    <Reference Include="Azure.Identity" />
+  </ItemGroup>
+
+  <ItemGroup Condition="'$(AspNetCoreMajorMinorVersion)' == '3.1'">
```
Comment (pakrym) [inline src/DataProtection/Azure.Storage.Blob/src/Microsoft.AspNetCore.DataProtection.Azure.Storage.Blob.csproj]: I don't think this one is required.

File: src/DataProtection/Azure.Storage.Blob/test/MockResponse.cs [hunk 11]
```diff
@@ -0,0 +1,90 @@
...
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Text;
+using Azure;
+using Azure.Core;
+
+namespace Microsoft.AspNetCore.DataProtection.Azure.Storage.Blob.Test
```
Comment (pranavkm) [inline src/DataProtection/Azure.Storage.Blob/test/MockResponse.cs]:
```suggestion
namespace Microsoft.AspNetCore.DataProtection.Azure.Storage.Blob
```

File: src/DataProtection/Azure.Storage.Blob/src/AzureDataProtectionBuilderExtensions.cs [hunk 12]
```diff
@@ -0,0 +1,123 @@
...
+using Microsoft.AspNetCore.DataProtection.KeyManagement;
+using Microsoft.Extensions.DependencyInjection;
+
+namespace Microsoft.AspNetCore.DataProtection
+{
+    /// <summary>
+    /// Contains Azure-specific extension methods for modifying a
+    /// <see cref="IDataProtectionBuilder"/>.
+    /// </summary>
+    public static class AzureDataProtectionBuilderExtensions
```
Comment (pranavkm) [inline src/DataProtection/Azure.Storage.Blob/src/AzureDataProtectionBuilderExtensions.cs]: Should this be `AzureStorageBlobDataProtectionBuilderExtensions`?

File: THIRD-PARTY-NOTICES.txt [hunk 13]
```diff
@@ -189,3 +189,27 @@ AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 THE SOFTWARE.
+
+License notice for azure-sdk-for-net
```
Comment (BrennanConroy) [inline THIRD-PARTY-NOTICES.txt]: cc @Pilchie

**Review by analogrelay — APPROVED** [review]

File: src/DataProtection/Azure.Storage.Blob/src/Microsoft.AspNetCore.DataProtection.Azure.Storage.Blob.csproj [hunk 14]
```diff
@@ -0,0 +1,21 @@
...
+
+  <PropertyGroup>
+    <Description>Microsoft Azure Blob storage support as key store.</Description>
+    <TargetFramework>netstandard2.0</TargetFramework>
+    <AllowUnsafeBlocks>true</AllowUnsafeBlocks>
+    <GenerateDocumentationFile>true</GenerateDocumentationFile>
+    <PackageTags>aspnetcore;dataprotection;azure;blob</PackageTags>
+    <IsShippingPackage>true</IsShippingPackage>
+    <UseLatestPackageReferences>true</UseLatestPackageReferences>
+    <PreReleaseVersionLabel>$(AzureSDKPreReleaseVersionLabel)</PreReleaseVersionLabel>
```
Comment (anurse) [inline src/DataProtection/Azure.Storage.Blob/src/Microsoft.AspNetCore.DataProtection.Azure.Storage.Blob.csproj]: I don't think this is working. Looking at the public CI artifacts (which perhaps are wrong) I see: `3.1.1-ci`. That indicates two things to me:

1. The `.1` patch version is applied. We want these to be `.0`.
1. The `-preview1` label is being lost, though again, perhaps that's a problem in the public CI.

@dotnet/aspnet-build do you have context on this? The same problem likely applies to the other two PRs (https://github.com/dotnet/aspnetcore/pull/17041, https://github.com/dotnet/extensions/pull/2618)

Comment (BrennanConroy) [reply src/DataProtection/Azure.Storage.Blob/src/Microsoft.AspNetCore.DataProtection.Azure.Storage.Blob.csproj]: The `-ci` is expected I believe, and `-preview1` should be applied in official builds. If you look at the other already shipped packages, you'll see they don't have `-ci`.

**Review by pakrym — APPROVED** [review]
LGTM!

File: src/DataProtection/Azure.Storage.Blob/src/Microsoft.AspNetCore.DataProtection.Azure.Storage.Blob.csproj (hunk 14 above)
Comment (dougbu) [reply src/DataProtection/Azure.Storage.Blob/src/Microsoft.AspNetCore.DataProtection.Azure.Storage.Blob.csproj]: The setup looks correct w.r.t. `$(PreReleaseVersionLabel)` and `$(DotNetFinalVersionKind)`.

> We want these to be `3.1.0`

Need to change the `$(VersionPrefix)` for that. I suggest the following in this project
```xml
<VersionPrefix>$(AspNetCoreMajorVersion).$(AspNetCoreMinorVersion).0</VersionPrefix>
```

Comment (anurse) [reply src/DataProtection/Azure.Storage.Blob/src/Microsoft.AspNetCore.DataProtection.Azure.Storage.Blob.csproj]: Sounds good. To be clear, this weird versioning will only hold true until we release the `.0`. Once that's done, we can just clean all the up and version them ahead to `3.1.x` (whatever the next `x` is). I'll make a note to track that for follow up after these packages go out.

Comment (BrennanConroy) [reply src/DataProtection/Azure.Storage.Blob/src/Microsoft.AspNetCore.DataProtection.Azure.Storage.Blob.csproj]: @dougbu Do you know how to do the same thing for Extensions? Setting VersionPrefix doesn't seem to affect the nupkg version

Comment (dougbu) [reply src/DataProtection/Azure.Storage.Blob/src/Microsoft.AspNetCore.DataProtection.Azure.Storage.Blob.csproj]: I believe there's an Arcade or .NET Core SDK setting that causes it to override `$(VerisonPrefix)`. IIRC that's set in dotnet/aspnetcore but not dotnet/extensions. Might be `$(IncludeSourceRevisionInInformationalVersion)`.

I didn't check a 3.x Arcade SDK but their logic in 'master' for calculating `$(VersionPrefix)` is in tools\Version.BeforeCommonTargets.targets and seems to no-op when the property is already set and `$(PrereleaseVersionLabel)` is non-empty.

Bottom line, no, I'm not sure and suggest you complete the research I started 😈

**Comment by anurse** [comment]
Packages look better now (`3.1.0-ci`, and I'm presuming the `-ci` will become `-preview1` in the official builds automatically).

**Comment by anurse** [comment]
Approved for March (since this is a bundle along with https://github.com/dotnet/extensions/pull/2618#issuecomment-575293740)

File: src/DataProtection/Azure.Storage.Blob/src/Microsoft.AspNetCore.DataProtection.Azure.Storage.Blob.csproj [hunk 15]
```diff
@@ -0,0 +1,22 @@
...
+    <DotNetFinalVersionKind />
+    <VersionPrefix>$(AspNetCoreMajorVersion).$(AspNetCoreMinorVersion).0</VersionPrefix>
```
Comment (BrennanConroy) [inline src/DataProtection/Azure.Storage.Blob/src/Microsoft.AspNetCore.DataProtection.Azure.Storage.Blob.csproj]:
```suggestion
```

We've decided to leave the version alone, so it'll have a patch version for the first release.

**Comment by anurse** [comment]
Closing as these changes are moving.
